File: sources/entity/tests/test_yahoo_finance.php
<?php
/**
 * Test Google Intelligence + Yahoo Finance + LinkedIn integration.
 * Run: php tests/test_yahoo_finance.php
 */

$passed = 0;
$failed = 0;

function test(string $label, $expected, $actual): void
{
    global $passed, $failed;
    if ($expected === $actual) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label}\n    Expected: " . json_encode($expected) . "\n    Got:      " . json_encode($actual) . "\n";
        $failed++;
    }
}

require_once __DIR__ . '/../php/tools.php';

$config = require __DIR__ . '/../php/config.php';
$tools = new LookupTools($config);

echo "=== Google Intelligence + Yahoo Finance ===\n\n";

// 1. Test Google Intelligence batch search
echo "1. Google Intelligence (3 searches in 1 API call):\n";
$intel = $tools->googleIntelligence('franchisebrands.co.uk');
test('Returns array', true, is_array($intel));
test('Has google_results key', true, isset($intel['google_results']));
test('Has yahoo_ticker key', true, array_key_exists('yahoo_ticker', $intel));
test('Has linkedin_url key', true, array_key_exists('linkedin_url', $intel));
test('Google results not empty', true, strlen($intel['google_results']) > 50);
test('Yahoo ticker is FRAN.L', 'FRAN.L', $intel['yahoo_ticker']);
test('LinkedIn URL found', true, $intel['linkedin_url'] !== null);
test('LinkedIn URL contains /company/', true, str_contains($intel['linkedin_url'] ?? '', '/company/'));

// 2. Test Yahoo Finance data fetch
echo "\n2. Yahoo Finance data fetch (FRAN.L):\n";
$data = $tools->yahooFinanceData('FRAN.L');
test('FRAN.L data not empty', true, strlen($data) > 100);
test('Contains company profile', true, str_contains($data, 'Company Profile'));
test('Contains Income Statement', true, str_contains($data, 'Income Statement'));
test('Contains Revenue', true, str_contains($data, 'Revenue'));
test('Contains source link', true, str_contains($data, 'finance.yahoo.com/quote/FRAN.L'));
test('Contains sector', true, str_contains($data, 'Sector:'));

// 3. Test LinkedIn company data fetch
echo "\n3. LinkedIn company data fetch:\n";
if ($intel['linkedin_url']) {
    $linkedinData = $tools->fetchLinkedInCompany($intel['linkedin_url']);
    test('LinkedIn data not empty', true, strlen($linkedinData) > 50);
    test('Contains company name', true, str_contains($linkedinData, 'Name:'));
    test('Contains LinkedIn Profile header', true, str_contains($linkedinData, 'LinkedIn Company Profile'));
    test('Contains address', true, str_contains($linkedinData, 'Address:'));
} else {
    echo "  SKIP: No LinkedIn URL found\n";
}

// 4. Test value formatting
echo "\n4. Value formatting:\n";
$ref = new ReflectionClass($tools);
$method = $ref->getMethod('yahooFormatVal');
test('Trillions', '50.7T', $method->invoke($tools, ['raw' => 50684952000000]));
test('Billions', '1.5B', $method->invoke($tools, ['raw' => 1500000000]));
test('Millions', '89.5M', $method->invoke($tools, ['raw' => 89460000]));
test('Thousands', '5.0K', $method->invoke($tools, ['raw' => 5000]));
test('Negative', '-1.5B', $method->invoke($tools, ['raw' => -1500000000]));
test('Null', '—', $method->invoke($tools, []));

echo "\n=== Results: {$passed} passed, {$failed} failed ===\n";
exit($failed > 0 ? 1 : 0);

File: sources/entity/app/tests/test_yahoo_finance.py
"""
Faithful port of ../../tests/test_yahoo_finance.php.

Google Intelligence + Yahoo Finance + LinkedIn integration.

PHP method mapping:
  googleIntelligence   -> google_intelligence(domain)
  yahooFinanceData     -> yahoo_finance_data(ticker)
  fetchLinkedInCompany -> fetch_linkedin_company(url)
  yahooFormatVal       -> _yahoo_format_val(field)   (PHP used reflection on the private)

The value-formatting block (PHP section 4) is PURE (no network) and runs by default.
Everything else hits Bright Data / Yahoo and is gated behind RUN_LIVE=1.
"""
import os

import pytest

from config import load_config
from tools import LookupTools


@pytest.fixture(scope="module")
def tools():
    return LookupTools(load_config())


live = pytest.mark.skipif(not os.environ.get("RUN_LIVE"), reason="live network")


# ══ 1. Google Intelligence (batch SERP) ═════════════════════════════════════

@pytest.mark.live
@live
def test_google_intelligence(tools):
    intel = tools.google_intelligence("franchisebrands.co.uk")
    assert isinstance(intel, dict)                          # Returns array
    assert "google_results" in intel                        # Has google_results key
    assert "yahoo_ticker" in intel                          # Has yahoo_ticker key
    assert "linkedin_url" in intel                          # Has linkedin_url key
    assert len(intel["google_results"]) > 50                # Google results not empty
    assert intel["yahoo_ticker"] == "FRAN.L"                # Yahoo ticker is FRAN.L
    assert intel["linkedin_url"] is not None                # LinkedIn URL found
    assert "/company/" in (intel["linkedin_url"] or "")     # LinkedIn URL contains /company/


# ══ 2. Yahoo Finance data fetch (FRAN.L) ════════════════════════════════════

@pytest.mark.live
@live
def test_yahoo_finance_data(tools):
    data = tools.yahoo_finance_data("FRAN.L")
    assert len(data) > 100                                  # FRAN.L data not empty
    assert "Company Profile" in data                        # Contains company profile
    assert "Income Statement" in data                       # Contains Income Statement
    assert "Revenue" in data                                # Contains Revenue
    assert "finance.yahoo.com/quote/FRAN.L" in data         # Contains source link
    assert "Sector:" in data                                # Contains sector


# ══ 3. LinkedIn company data fetch ══════════════════════════════════════════

@pytest.mark.live
@live
def test_fetch_linkedin_company(tools):
    intel = tools.google_intelligence("franchisebrands.co.uk")
    if not intel["linkedin_url"]:
        pytest.skip("No LinkedIn URL found")
    linkedin = tools.fetch_linkedin_company(intel["linkedin_url"])
    assert len(linkedin) > 50                               # LinkedIn data not empty
    assert "Name:" in linkedin                              # Contains company name
    assert "LinkedIn Company Profile" in linkedin           # Contains header
    assert "Address:" in linkedin                           # Contains address


# ══ 4. Value formatting (PURE — no network) ═════════════════════════════════

def test_yahoo_format_trillions(tools):
    assert tools._yahoo_format_val({"raw": 50684952000000}) == "50.7T"


def test_yahoo_format_billions(tools):
    assert tools._yahoo_format_val({"raw": 1500000000}) == "1.5B"


def test_yahoo_format_millions(tools):
    assert tools._yahoo_format_val({"raw": 89460000}) == "89.5M"


def test_yahoo_format_thousands(tools):
    assert tools._yahoo_format_val({"raw": 5000}) == "5.0K"


def test_yahoo_format_negative(tools):
    assert tools._yahoo_format_val({"raw": -1500000000}) == "-1.5B"


def test_yahoo_format_null(tools):
    assert tools._yahoo_format_val({}) == "—"
